spark/src/main/scala/org/apache/spark/sql/delta/schema/ImplicitMetadataOperation.scala

```diff
@@ -227,6 +227,149 @@ object ImplicitMetadataOperation {
+      spark, path, protocol, metadata.schema)
+    if (dependentGeneratedColumns.nonEmpty) {
+      throw DeltaErrors.generatedColumnsDataTypeMismatch(
+        path,
+        currentDt,
+        updateDt,
+        dependentGeneratedColumns
+      )
+    }
+  }
+
+  /**
+   * Check whether the provided field is currently being referenced
+   * by CHECK constraints or generated columns.
+   * Note that we explicitly ignore the check for `StructType` in this
+   * function, since any `StructType` will be checked in
+   * [[checkReferencedByCheckConstraintsOrGeneratedColumns]].
+   *
+   * @param spark the spark session used.
+   * @param path the full column path for the current field.
+   * @param protocol the protocol used.
+   * @param metadata the metadata used for checking constraints and generated columns.
+   * @param currentDt the current data type.
+   * @param updateDt the updated data type.
+   */
+  private def checkConstraintsOrGeneratedColumnsExceptStruct(
+      spark: SparkSession,
+      path: Seq[String],
+      protocol: Protocol,
+      metadata: Metadata,
+      currentDt: DataType,
+      updateDt: DataType): Unit = (currentDt, updateDt) match {
+    case (StructType(_), StructType(_)) =>
+      // we explicitly ignore the check for `StructType` here.
+    case (from, to) =>
+      if (currentDt != updateDt) {
+        checkDependentConstraints(spark, path, metadata, from, to)
+        checkDependentGeneratedColumns(spark, path, protocol, metadata, from, to)
+      }
+  }
+
+  /**
+   * If the provided data types are `StructType`, check whether the nested/inner fields
+   * of the struct are being referenced by CHECK constraints or generated columns.
+   * It is worth noting that:
+   * 1. we only care about the inner fields of `StructType`,
+   *    for all the other types, e.g., `MapType` or `ArrayType`, we will directly check
+   *    the dependant constraints and generated columns without inspecting the inner fields;
+   *    this indicates we intentionally keep being too strict for `MapType` and `ArrayType`.
+   *
+   * 2. this function will not recursively check any nested `StructType`, the actual traverse
+   *    through the nested struct(s) will be handled by [[SchemaMergingUtils.transformColumns]]
+   *    in [[checkDependentExpressions]].
+   *
+   * @param spark the spark session used.
+   * @param protocol the protocol used.
+   * @param path the full column path for the current field.
+   * @param from the current data type.
+   * @param to the updated data type.
+   * @param metadata the metadata used for checking constraints and generated columns.
+   */
+  private def checkReferencedByCheckConstraintsOrGeneratedColumns(
+      spark: SparkSession,
+      protocol: Protocol,
+      path: Seq[String],
+      from: DataType,
+      to: DataType,
+      metadata: Metadata): Unit = (from, to) match {
+    case (StructType(fromFields), StructType(toFields)) =>
```

Comment:
I think you may be able to simplify a bit by removing this layer (`checkReferencedByCheckConstraintsOrGeneratedColumns`) and call `checkConstraintsOrGeneratedColumnsExceptStruct` directly from `checkDependentExpressions`, since `transformColumns` will call that method on all struct fields in the schema already:


```
  private def checkConstraintsOrGeneratedColumnsExceptStruct(
      spark: SparkSession,
      path: Seq[String],
      protocol: Protocol,
      metadata: Metadata,
      currentDt: DataType,
      updateDt: DataType): Unit = (currentDt, updateDt) match {
    case (StructType(_), StructType(_)) =>
      // we explicitly ignore the check for `StructType` here.
    case (from, to) =>
      if (currentDt != updateDt) {
        checkDependentConstraints(spark, path, metadata, from, to)
        checkDependentGeneratedColumns(spark, path, protocol, metadata, from, to)
      }
  }

    SchemaMergingUtils.transformColumns(metadata.schema, dataSchema) {
      case (fieldPath, currentField, Some(updateField), _)
        if !SchemaMergingUtils.equalsIgnoreCaseAndCompatibleNullability(
          currentField.dataType,
          updateField.dataType
        ) =>
        checkConstraintsOrGeneratedColumnsExceptStruct(
          spark = sparkSession,
          protocol = protocol,
          path = fieldPath :+ currentField.name,
          from = currentField.dataType,
          to = updateField.dataType,
          metadata = metadata
        )
```

nit: I would rename `checkConstraintsOrGeneratedColumnsExceptStruct` to e.g. `checkDependentExpressionsOnStructField`

Reply:
thanks for the suggestion!
- `checkReferencedByCheckConstraintsOrGeneratedColumns` has been removed.
- `checkDependentExpressionsOnStructField` is now directly called in `checkDependentExpressions`.